linked_list/single_linked_list.c

```diff
@@ -0,0 +1,186 @@
+#include <stdio.h>
+#include <stdlib.h>
+#define ASSERT(x) printf("Assertion: " #x " = %d\n", (x))
+
+typedef int data_t;
+
+typedef struct node {
+  data_t data;
+  struct node *next;
+} node;
+
+#define CHECK_ALLOC(ptr)						\
+  if (!ptr) {								\
+    printf("Error allocating memory for %s in  %s: %d line", #ptr, __func__, __LINE__-1); \
+    exit(2);								\
+  }
+
+node* list_create(data_t data){
+  node* first = (node*)malloc(sizeof(node));
+  CHECK_ALLOC(first);
+  first->data = data;
+  first->next = NULL;
+  return first;
+}
+
+
+void list_free(node* first){
+  while (first != NULL)
+    {
+      node* tmp = first;
+      first = first->next;
+      free(tmp);
+    }
+}
+
+void list_append_element_end(node* list, int new_element){
+  node* new_node = (node*)malloc(sizeof(node)); 
+  CHECK_ALLOC(new_node); 
+  new_node->data = new_element;
+  new_node->next = NULL;
+
+  if (list == NULL){
+    list = new_node;
+  }
+  while (list->next != NULL)
+    list = list->next;
+
+  list->next = new_node;
+}
```

Comment:
1) `list_append_element_end` is pretty long as for a function (sometimes very long function names are okay, e.g. for test functions) - I would name it just `list_append`, as `append` is known to add to an end (e.g. `.append()` method in Python).
2) This won't work when `list == NULL`. Add such a test and try to find out what's wrong:
```
void test_append_to_empty() {
    node* head = NULL;
    list_append_element_end(head, 1);
    ASSERT(list != NULL);
    ASSERT(list->data == 1);
    ASSERT(list->next == NULL);
    list_free(head);
}
```